float32[] cp_pos
---
float32[] tf_pos


